cassandra/timestamps.py

```diff
@@ -0,0 +1,107 @@
+# limitations under the License.
+
+"""
+This module contains utilities for generating timestamps for client-side
+timestamp specification.
+"""
+
+import logging
+import time
+from threading import Lock
+
+log = logging.getLogger(__name__)
+
+class MonotonicTimestampGenerator(object):
+    """
+    An object that, when called, returns ``int(time.time() * 1e6)`` when
+    possible, but, if the value returned by ``time.time`` doesn't increase,
+    drifts into the future and logs warnings.
+    Exposed configuration attributes can be configured with arguments to
+    ``__init__`` or by changing attributes on an initialized object.
+
+    .. versionadded:: 3.8.0
+    """
+
+    warn_on_drift = True
+    """
+    If true, log warnings when timestamps drift into the future as allowed by
+    :attr:`warning_threshold` and :attr:`warning_interval`.
+    """
+
+    warning_threshold = 0
+    """
+    This object will only issue warnings when the returned timestamp drifts
+    more than ``warning_threshold`` seconds into the future.
+    """
+
+    warning_interval = 0
+    """
+    This object will only issue warnings every ``warning_interval`` seconds.
+    """
+
+    def __init__(self, warn_on_drift=True, warning_threshold=0, warning_interval=0):
+        self.lock = Lock()
+        with self.lock:
+            self.last = 0
+            self._last_warn = 0
+        self.warn_on_drift = warn_on_drift
+        self.warning_threshold = warning_threshold
+        self.warning_interval = warning_interval
+
+    def _next_timestamp(self, now, last):
+        """
+        Returns the timestamp that should be used if ``now`` is the current
+        time and ``last`` is the last timestamp returned by this object.
+        Intended for internal and testing use only; to generate timestamps,
+        call an instantiated ``MonotonicTimestampGenerator`` object.
+
+        :param int now: an integer to be used as the current time, typically
+            representing the current time in seconds since the UNIX epoch
+        :param int last: an integer representing the last timestamp returned by
+            this object
+        """
+        if now > last:
+            self.last = now
+            return now
+        else:
+            self._maybe_warn(now=now)
+            self.last = last + 1
+            return self.last
+
+    def __call__(self):
+        """
+        Makes ``MonotonicTimestampGenerator`` objects callable; defers
+        internally to _next_timestamp.
+        """
+        with self.lock:
+            return self._next_timestamp(now=int(time.time() * 1e6),
+                                        last=self.last)
+
+    def _maybe_warn(self, now):
+        # should be called from inside the self.lock.
+        diff = now - self._last_warn
+        since_last_warn = now - self._last_warn
+
+        warn = (self.warn_on_drift and
+                (diff >= self.warning_threshold) and
```

Comment:
I think it should be `since_last_warn >= self.warning_threshold` -- we're trying to avoid flooding logs, so we care about how long it's been since the last warning was issued. Sound right?